Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm

```diff
@@ -10973,19 +10973,37 @@ - (void)_completeImageAnalysisRequestForContextMenu:(CGImageRef)image requestIde
     auto weakSelf = WeakObjCPtr<WKContentView>(self);
     auto aggregator = CallbackAggregator::create([weakSelf, data]() mutable {
         auto strongSelf = weakSelf.get();
-        if (!strongSelf || !strongSelf->_canUpdateVisibleContextMenuWithImageAnalysisActions)
+        if (!strongSelf || !strongSelf->_waitingForDynamicImageAnalysisContextMenuActions)
             return;
 
+        strongSelf->_waitingForDynamicImageAnalysisContextMenuActions = NO;
         strongSelf->_imageAnalysisContextMenuActionData = WTFMove(*data);
         [[strongSelf contextMenuInteraction] updateVisibleMenuWithBlock:makeBlockPtr([strongSelf](UIMenu *menu) -> UIMenu * {
+            __block auto indexOfPlaceholder = NSNotFound;
+            __block BOOL foundCopyItem = NO;
+            auto revealImageIdentifier = elementActionTypeToUIActionIdentifier(_WKElementActionTypeRevealImage);
+            auto copyIdentifier = elementActionTypeToUIActionIdentifier(_WKElementActionTypeCopy);
+            [menu.children enumerateObjectsUsingBlock:^(UIMenuElement *child, NSUInteger index, BOOL* stop) {
+                auto *action = dynamic_objc_cast<UIAction>(child);
+                if ([action.identifier isEqualToString:revealImageIdentifier] && (action.attributes & UIMenuElementAttributesHidden))
+                    indexOfPlaceholder = index;
+                else if ([action.identifier isEqualToString:copyIdentifier])
+                    foundCopyItem = YES;
+            }];
+
+            if (indexOfPlaceholder == NSNotFound)
+                return menu;
+
             auto adjustedChildren = adoptNS(menu.children.mutableCopy);
+            auto replacements = adoptNS([NSMutableArray<UIMenuElement *> new]);
+
             auto *elementInfo = [_WKActivatedElementInfo activatedElementInfoWithInteractionInformationAtPosition:strongSelf->_positionInformation userInfo:nil];
             auto addAction = [&](_WKElementActionType action) {
                 auto *elementAction = [_WKElementAction _elementActionWithType:action info:elementInfo assistant:strongSelf->_actionSheetAssistant.get()];
-                [adjustedChildren addObject:[elementAction uiActionForElementInfo:elementInfo]];
+                [replacements addObject:[elementAction uiActionForElementInfo:elementInfo]];
             };
 
-            if ([strongSelf copySubjectResultForImageContextMenu])
+            if (foundCopyItem && [strongSelf copySubjectResultForImageContextMenu])
```

Comment:
Is the idea here that we only want to show "Copy Subject" if we have some sort of general "Copy"?

Reply:
Indeed. I explained the rationale in the commit message, but to summarize — I think it's a sensible default behavior for WebKit clients, and in particular it prevents the Copy action from showing up in (enterprise) managed emails in Mail on iOS.